### PR TITLE
Custom Filters for Code Modules

### PR DESCRIPTION
**Pull request by devorama**
Add ability to add custom namespace filters to allow 
config of seperate code not to require change of core systems config
files. This will allow easier updates of core system without
concern of overwriting custom filters that was added.

I dont know how todo a Testcase for this if anybody has ideas I will do it

Also If accepted the Code Modules doc should be update, I have not done before but would like to
if it is accepted

Signed-off-by: devorama <33496470+devorama@users.noreply.github.com>

**Checklist:**
- [ ] Securely signed commits
- [ ] Component(s) with PHPdocs
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [ ] Conforms to style guide

### CODE REVIEW

**Comment by avegacms** [comment]
I think it's a good feature for modules

File: system/Filters/Filters.php
Comment (samsonasik) [inline system/Filters/Filters.php:114]: `$this->config` is a `\Config\Filters` instance, so, can use the following check:

```suggestion
              if (isset($this->config->discoverFilters) && $this->config->discoverFilters)
```

Comment (devorama) [reply system/Filters/Filters.php]: Thanks you, have made changes this makes alot more sense

**Comment by devorama** [comment]
> Great addition! PR looks pretty good, please be sure to run it through the code sniffer for style guide compliance. I would also 
like a test case to confirm discovery is working as intended.
@MGatner  Hi there, If you can give suggestions on the test case for that is the part that I dunno as all the test cases are based around hardcoded config being passed and not a actual Code module system running inside the actual whole CI framework, I will try and look at other test cases on how the  load but it seems all are more directly related to specific modules areas and not loading the whole code module, thus really stummped at writing a test for it. Personally have tested it in my own enviroment and knows it work but understand theneed of the test case for this.

**Comment by paulbalandan** [comment]
You should probably use `Config\Modules` in setting up auto-discovery just like the others, like events, etc.

**Comment by devorama** [comment]
@MGatner  and @paulbalandan  I think this solution for testing the custom filter is valid and will work. Basically I have added the custom filter under tests/_support/Config/Filters.php and it creates a class for it under tests/_support/Filters/Customfilter.php I then updated the routes to add a new route that uses this filter.  tests/_support/Config/Routes.php to look like this $routes->add('testingfilter', 'TestController::index',  ['filter' => 'test-customfilter']); normally as I remember this creates a error if the filter was not found, I will look to maybe add another test controller to be called regardless to make sure it gets called but I think this is the best way one could emulate a code module and test the actual use of the functionality in the system, hope that will be okay

**Comment by devorama** [comment]
trying to get routes and filters to load as a code module correctly when doing the codeignitertest  as CI's AutoloaderConfig does load the _support as a code module
prent

Found that routes file did not have namespace added that hopedully it will solve the probelm and route will point to hello controller directly for testing, hope this works

**Comment by michalsn** [comment]
Maybe I am missing something but is there any reason why this shouldn't be handled via `Config/Modules` as @paulbalandan pointed out?

**Comment by devorama** [comment]
> 
> 
> Maybe I am missing something but is there any reason why this shouldn't be handled via `Config/Modules` as @paulbalandan pointed out?

I basically dont know enough of CI inner workings yet to write a test for this, for looking at how events test work with Config/Modules I dont see how one can do that for Filters with it. I will look again but test wise I wanted to achieve a test to load it and use it, but I think the closes I can come maybe is to test the autodiscovery but it feels wrong as need to confirm it will work inside the workings between loading custom filter from code module and using in a router and I cant achieve that , but will re-look

**Comment by devorama** [comment]
I checked code again and Modules work for Events and routes for they actually have code for it in their classes and a module config get passed to them and then inside they basically use the module as the config flag I used in the filters config file. I did not want to change teh Filters constructor and dont think it is wise to just create a default module object (config('Modules') ) . Basically still stumped how to test this. If anybody can explain what I am missing will appreciate for the actual code works I know I am using it, but to write a test is confusing as I actually dont know how the test system works behind the scenes when and how is certain objects loaded, is the whole CI system loaded at a point etc, basically a few stuff for me still to learn before I can fully grasp how to get the test modules in place.

**Comment by devorama** [comment]
@MGatner @michalsn @paulbalandan  Thank you I figured out how the testing system now loads and how to manipulate it for testing purposes, the testing will now be successfull and test the loading and calling of a custom filter from a code module.  I was considering to introduce the Modules unit into filters also as with Events and Routes to allow a person to enable and disable the module as with those, this will require me to change the contructor of Filters by adding the module config at the end , I will default it to null and if it is null then load the default config('Modules') for example, this will then remove the need of the configuration variable in Filters config itself to set discovering of filters on or off. Must I rather do that for this instead of a new pull request ?

**Comment by devorama** [comment]
I made the change as it is more inline with the design of CI to use the modules config to enable disable the discovery of custom files from there, thus removed the new config value in filters config and added the logic to use module config as required

File: system/Filters/Filters.php
Comment (samsonasik) [inline system/Filters/Filters.php:115]: @param \Config\Modules|null

File: system/Filters/Filters.php
Comment (samsonasik) [inline system/Filters/Filters.php:126]: can use $this->moduleConfig = $moduleConfig ?? config('Modules');

File: system/Filters/Filters.php
Comment (samsonasik) [inline system/Filters/Filters.php:117]: need \Config\Modules type hint

**Comment by samsonasik** [comment]
please rebase latest develop, phpstan level increased to level 3 that can do more static analysis check

File: system/Filters/Filters.php
Comment (devorama) [reply system/Filters/Filters.php:117]: Hi ther, I am little confused how todo this, copied from RouterCollection so that I try and follow the standard and it looks like this , how do I do that Config type hint ?
prent

**Comment by devorama** [comment]
> 
> 
> please rebase latest develop, phpstan level increased to level 3 that can do more static analysis check

Hi there, I think have it correct  now ? it seems to be happy with the statis analysis check in all and looks good I think, no idea what it does.
prent

File: system/Filters/Filters.php
Comment (devorama) [reply system/Filters/Filters.php]: K I think I have booked a solution in that will work, I looked at the FileLocator in RouteCollection and copied the same concept from it for Modules.

**Comment by devorama** [comment]
> 
> 
> please rebase latest develop, phpstan level increased to level 3 that can do more static analysis check

Followed your tutorial thank you and rebased and forced update back in as per tutorial. should all be happy now

File: system/Filters/Filters.php
Comment (michalsn) [inline system/Filters/Filters.php:118]:
```suggestion
	public function __construct($config, RequestInterface $request, ResponseInterface $response, Modules $moduleConfig = null)
```

File: system/Filters/Filters.php
Comment (michalsn) [inline system/Filters/Filters.php:124]:
```suggestion
```

File: system/Filters/Filters.php
Comment (michalsn) [inline system/Filters/Filters.php:135]:
```suggestion
	 * $filters->aliases['custom-auth'] = \Acme\Blob\Filters\BlobAuth::class;
```

**Review by michalsn — CHANGES_REQUESTED** [review]
I think this is really the last round of changes from me. Thanks!

**Review by michalsn — APPROVED** [review]
It looks good to me, but I'll let @MGatner review this before merging as he was already involved here.

File: app/Config/Filters.php
Comment (MGatner) [inline app/Config/Filters.php:7]: Please remove this newline, as it constitutes a change to a file in **app** that otherwise does not need to be touched.
```suggestion
```

File: tests/_support/Config/Routes.php
Comment (MGatner) [inline tests/_support/Config/Routes.php:1]: Extra space here.
```suggestion
<?php namespace Tests\Support\Config;
```

File: tests/system/CodeIgniterTest.php
Comment (MGatner) [inline tests/system/CodeIgniterTest.php:59]: This file does not need any changes.
```suggestion
```

Comment (devorama) [reply tests/system/CodeIgniterTest.php:59]: Sorry was one of the places I tried todo a test for the filter from will correct

**Comment by devorama** [comment]
No idea why Static Analysis Check is giving this error, it ons files I did not ouch ?
prent

**Comment by samsonasik** [comment]
It seems new phpstan released https://github.com/phpstan/phpstan/releases/tag/0.12.40

You seems can remove `// @phpstan-ignore-line` at:

-  `system/HTTP/Response.php` line 902
- `system/I18n/Time.php` line 560

**Comment by samsonasik** [comment]
I created PR https://github.com/codeigniter4/CodeIgniter4/pull/3547 for phpstan notice

**Pull request merged**
